compiler/src/openxla/compiler/nvgpu/Transforms/NormalizeHLOConvolutionLayouts.cpp

```diff
@@ -0,0 +1,165 @@
+    if (lhsPermutation == identityPermuation &&
+        rhsPermutation == identityPermuation &&
+        resultPermutation == identityPermuation) {
+      return failure();
+    }
+
+    // Transpose tensor operand if needed.
+    Location loc = op.getLoc();
+    Value lhs = op.getLhs();
+    if (lhsPermutation != identityPermuation) {
+      lhs = rewriter.createOrFold<stablehlo::TransposeOp>(
+          loc, lhs, rewriter.getI64TensorAttr(lhsPermutation));
+    }
+
+    // Transpose kernel operand if needed.
+    Value rhs = op.getRhs();
+    if (rhsPermutation != identityPermuation) {
+      rhs = rewriter.createOrFold<stablehlo::TransposeOp>(
+          loc, rhs, rewriter.getI64TensorAttr(rhsPermutation));
+    }
+
+    // Derive new result type.
+    auto resultTy = op.getResult().getType().cast<RankedTensorType>();
+    auto shape = resultTy.getShape();
+    SmallVector<int64_t> newShape = {shape[opDims.getOutputBatchDimension()],
+                                     shape[outputSpatialDimensions[0]],
+                                     shape[outputSpatialDimensions[1]],
+                                     shape[opDims.getOutputFeatureDimension()]};
+    auto newResultTy =
+        RankedTensorType::get(newShape, resultTy.getElementType());
+
+    // Recreate convolution op.
+    auto newDimsAttr = mlir::stablehlo::ConvDimensionNumbersAttr::get(
+        op.getContext(),
+        /*inputBatchDimension=*/0,
+        /*inputFeatureDimension=*/3,
+        /*inputSpatialDimensions=*/{1, 2},
+        /*kernelInputFeatureDimension=*/1,
+        /*kernelOutputFeatureDimension=*/0,
+        /*kernelSpatialDimensions=*/{2, 3},
+        /*outputBatchDimension=*/0,
+        /*outputFeatureDimension=*/3,
+        /*outputSpatialDimension=*/{1, 2});
+    auto newOp = rewriter.create<stablehlo::ConvolutionOp>(
+        loc, newResultTy, lhs, rhs, op.getWindowStridesAttr(),
+        op.getPaddingAttr(), op.getLhsDilationAttr(), op.getRhsDilationAttr(),
+        op.getWindowReversalAttr(), newDimsAttr, op.getFeatureGroupCountAttr(),
+        op.getBatchGroupCountAttr(), op.getPrecisionConfigAttr());
+
+    // Transpose result if needed.
+    Value result = newOp.getResult();
+    if (resultPermutation != identityPermuation) {
+      result = rewriter.createOrFold<stablehlo::TransposeOp>(
+          loc, result, rewriter.getI64TensorAttr(resultPermutation));
+    }
+
+    rewriter.replaceOp(op, result);
+    return success();
+  }
+
+  Layout tensorLayout, kernelLayout;
+};
+
+class NormalizeHLOConvolutionLayoutsPass
+    : public ::impl::NormalizeHLOConvolutionLayoutsPassBase<
+          NormalizeHLOConvolutionLayoutsPass> {
+ public:
+  void runOnOperation() override {
+    MLIRContext* ctx = &getContext();
+
+    // Populate patterns.
+    RewritePatternSet patterns(ctx);
+    patterns.add<NormalizeConvolutionLayoutPattern>(
+        ctx, /*tensorLayout=*/Layout::NHWC, /*kernelLayout*/ Layout::KCHW);
```

Comment:
Would it make sense to turn these into options? I'm really just asking, it might be more trouble than worth it.